sound/soc/sof/ipc.c

```diff
@@ -463,20 +450,34 @@ void snd_sof_ipc_msgs_rx(struct snd_sof_dev *sdev)
 	case SOF_IPC_GLB_PM_MSG:
 		break;
 	case SOF_IPC_GLB_COMP_MSG:
-		ipc_comp_notification(sdev, &hdr);
+		rx_callback = ipc_comp_notification;
 		break;
 	case SOF_IPC_GLB_STREAM_MSG:
-		/* need to pass msg id into the function */
-		ipc_stream_message(sdev, hdr.cmd);
+		rx_callback = ipc_stream_message;
 		break;
 	case SOF_IPC_GLB_TRACE_MSG:
-		ipc_trace_message(sdev, type);
+		rx_callback = ipc_trace_message;
 		break;
 	default:
 		dev_err(sdev->dev, "error: unknown DSP message 0x%x\n", cmd);
 		break;
 	}
 
+	if (rx_callback) {
+		/* read the full message as there we have rx handler for it */
+		full_msg = kzalloc(hdr.size, GFP_KERNEL);
```

Comment:
does it have to be zeroed? Wouldn't it be completely overwritten anyway?

Reply:
True, there is another kzalloc which can be changed when registering a notification request, if list_add() is still happy.